### DISCUSSION TITLE
A problem with different thermal conductivities: a general question about the implementation

### DISCUSSION THREAD

**Discussion by amnchacon** [firedrakeproject/firedrake#3396] [Firedrake support]
Hello everyone. Imagine that I have a wall composed of different materials, so we have different thermal conductivities. Is it possible to implement a heat conduction problem in Firedrake with this fact? because those thermal conductivities would affect our stiffness matrix $\int_\Omega k \nabla \theta \nabla v \hspace{1mm} d\Omega$ ....

![Captura de tela de 2024-02-07 16-45-01](https://github.com/firedrakeproject/firedrake/assets/146268309/ff1d71ef-590b-4a8b-a11c-0a77df5009ec)

**Comment by colinjcotter** [comment]
This is a maths question rather than a Firedrake question. But you just integrate by parts to derive the formula in the usual way.

On 7 Feb 2024, at 22:49, Andrés Mauricio Nieves Chacón ***@***.***> wrote:

﻿

Hello everyone. Imagine that I have a wall composed of different materials, so we have different thermal conductivities. Is it possible to implement a heat conduction issue in Firedrake with this fact? because those thermal conductivities would affect our stiffness matrix $\int_\Omega k \nabla \theta \nabla v \hspace{1mm} d\Omega$ ....

Captura.de.tela.de.2024-02-07.16-45-01.png (view on web)<https://github.com/firedrakeproject/firedrake/assets/146268309/ff1d71ef-590b-4a8b-a11c-0a77df5009ec>

**Reply by amnchacon** [reply]
But I would have to denote which parts of my mesh will have an specific thermal conductivity, how to specify that in Firedrake? how to translate it in the variational formulation? if I have a Unit Interval Mesh, how could I specify that the first thermal conductivity is applied from one specific point to another, and the same with follow thermal conductivy... and the follow, and that way sequentially. If I have a square mesh for example, where the changes of thermal conductivity are not only on the boundaries but within the domain, how would I specify it? is possible specify the integration limits within the variational formulation written in Firedrake, where govern an specific thermal conductivity?

**Reply by ksagiyam** [reply]
Details depend on your math problem, but the followings probably help.

You could use `conditional` to represent thermal conductivity; see, e.g., https://github.com/firedrakeproject/firedrake/blob/master/tests/regression/test_conditional.py for examples.
If the surface of discontinuity coincides with the cell boundaries, you could interpolate the thermal conductivity into the DG0 space.

You could specify subdomain of cell integration as `dx(subdomain_label)` just as you do for `ds`. If subdomains are not labeled, you could label them using `mark_entities()` method of the mesh object: https://www.firedrakeproject.org/firedrake.html#firedrake.mesh.MeshGeometry.mark_entities.

**Reply by amnchacon** [reply]
I see. That's what I needed to know. I'll check it out carefully. Thanks for this information Mr.
